modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/snapshot/SnapshotCheckProcess.java

```diff
@@ -0,0 +1,543 @@
+     * @param rqFilter If not {@code null}, used to filter which requests/process to stop. If {@code null}, stops all the validations.
+     */
+    public void interrupt(Throwable th, @Nullable Function<SnapshotCheckProcessRequest, Boolean> rqFilter) {
+        requests.values().forEach(rq -> {
+            if (rqFilter == null || rqFilter.apply(rq)) {
+                rq.error(th);
+
+                clean(rq.requestId(), th, null, null);
+            }
+        });
+    }
+
+    /** Stops the related validation if the node is a mandatory one. */
+    private void nodeLeft(ClusterNode node) {
+        if (node.isClient() || requests.isEmpty())
+            return;
+
+        interrupt(
+            new ClusterTopologyCheckedException("Snapshot checking stopped. A node left the cluster: " + node + '.'),
+            rq -> rq.nodes().contains(node.id())
+        );
+    }
+
+    /** Phase 2 and process finish. */
+    private IgniteInternalFuture<?> reduceValidatePartsAndFinish(
+        UUID procId,
+        Map<UUID, HashMap<PartitionKeyV2, PartitionHashRecordV2>> results,
+        Map<UUID, Throwable> errors
+    ) {
+        clean(procId, null, results, errors);
+
+        return new GridFinishedFuture<>();
+    }
+
+    /** Phase 2 beginning.  */
+    private IgniteInternalFuture<HashMap<PartitionKeyV2, PartitionHashRecordV2>> validateParts(SnapshotCheckProcessRequest req) {
+        SnapshotCheckProcessRequest locReq;
+
+        if (stopAndCleanOnError(req, null) || (locReq = requests.get(req.snapshotName())) == null)
+            return new GridFinishedFuture<>();
+
+        assert locReq.equals(req);
+
+        // Store metas to collect cluster operation result later.
+        GridFutureAdapter<SnapshotPartitionsVerifyTaskResult> clusterOpFut = clusterOpFuts.get(req.requestId());
+
+        if (clusterOpFut != null)
+            locReq.metas = req.metas;
+
+        // Local meta might be null if current node started after the snapshot creation or placement.
+        if (!req.nodes.contains(kctx.localNodeId()) || locReq.meta() == null)
+            return new GridFinishedFuture<>();
+
+        GridFutureAdapter<HashMap<PartitionKeyV2, PartitionHashRecordV2>> locPartsChkFut = new GridFutureAdapter<>();
+
+        locReq.fut(locPartsChkFut);
+
+        ExecutorService executor = kctx.cache().context().snapshotMgr().snapshotExecutorService();
+
+        executor.submit(() -> stopFutureOnAnyFailure(locPartsChkFut, () -> {
+            // An error can occure when the local future is still null.
+            if (locReq.error() != null)
+                locPartsChkFut.onDone(locReq.error());
+
+            if (locPartsChkFut.isDone())
+                return;
+
+            File snpDir = kctx.cache().context().snapshotMgr().snapshotLocalDir(locReq.snapshotName(), locReq.snapshotPath());
+
+            try {
+                Map<PartitionKeyV2, PartitionHashRecordV2> res = kctx.cache().context().snapshotMgr().checker()
```

Comment:
Looks weird that `SnapshotCheckProcess` doesn't have own instance of `SnapshotChecker`. `IgniteSnapshotManager` checks nothing but uses only `read*` methods from the `SnapshotChecker` class. Consider refactoring here.

Reply:
I thought about it. The checker is shared. The tasks use it too. Once we remove the tasks, we could move the checker to the process. WDYT?